### PR TITLE
ZESC-2120 Upgrade PHP to 7.4.27

### PR DESCRIPTION
**Pull request by jholder85638**
https://jira.corp.synacor.com/browse/ZBUG-2498
Upgrade PHP to resolve:
5 CVEs since previous version of 7.3.25 with a max CVSS Score of 6.4
CVE-2019-11050 score: 6.4 (Not really vulnerable, we don't use PHP EXIF extension)
CVE-2019-11047 score: 6.4 (Not really vulnerable, we don't use PHP EXIF extension)
CVE-2019-11046 score: 5.0
CVE-2019-11045 score: 4.3
CVE-2019-11044 score: 5.0
```
CVE ID,Score,Desc
CVE-2019-11050,6.4,When PHP EXIF extension is parsing EXIF information from an image, e.g. via exif_read_data() function, in PHP versions 7.2.x below 7.2.26, 7.3.x below 7.3.13 and 7.4.0 it is possible to supply it with data what will cause it to read past the allocated buffer. This may lead to information disclosure or crash.
CVE-2019-11047,6.4,When PHP EXIF extension is parsing EXIF information from an image, e.g. via exif_read_data() function, in PHP versions 7.2.x below 7.2.26, 7.3.x below 7.3.13 and 7.4.0 it is possible to supply it with data what will cause it to read past the allocated buffer. This may lead to information disclosure or crash.
CVE-2019-11046,5.0,In PHP versions 7.2.x below 7.2.26, 7.3.x below 7.3.13 and 7.4.0, PHP bcmath extension functions on some systems, including Windows, can be tricked into reading beyond the allocated space by supplying it with string containing characters that are identified as numeric by the OS but aren't ASCII numbers. This can read to disclosure of the content of some memory locations.
CVE-2019-11045,4.3,In PHP versions 7.2.x below 7.2.26, 7.3.x below 7.3.13 and 7.4.0, PHP DirectoryIterator class accepts filenames with embedded \0 byte and treats them as terminating at that byte. This could lead to security vulnerabilities, e.g. in applications checking paths that the code is allowed to access.
CVE-2019-11044,5.0,In PHP versions 7.2.x below 7.2.26, 7.3.x below 7.3.13 and 7.4.0 on Windows, PHP link() function accepts filenames with embedded \0 byte and treats them as terminating at that byte. This could lead to security vulnerabilities, e.g. in applications checking paths that the code is allowed to access.
```
Note: Zimbra was never vulnerable to CVE-2006-3017 CVE score 9.3: https://www.cvedetails.com/cve/CVE-2006-3017/

### CODE REVIEW

File: thirdparty/php/zimbra-php/debian/changelog [hunk 1]
```diff
@@ -1,3 +1,9 @@
+zimbra-php (VERSION-1zimbra8.7b3ZAPPEND) unstable; urgency=medium
+
+  * Upgraded PHP to 7.4.27 (ZBUG-2498)
+
+   -- Zimbra Packaging Services <packaging-devel@zimbra.com>  Mon, 14 Feb 2022 21:56:52 +0000
```
Comment (k-kato) [inline thirdparty/php/zimbra-php/debian/changelog:5]: nit: fix indent. Use a single white space as same as other "-- xxx" descriptions.
```suggestion
 -- Zimbra Packaging Services <packaging-devel@zimbra.com>  Mon, 14 Feb 2022 21:56:52 +0000
```

File: thirdparty/php/zimbra-php/rpm/SPECS/php.spec [hunk 2]
```diff
@@ -19,6 +19,8 @@ URL:                http://php.net
 The Zimbra PHP build
 
 %changelog
+* Mon 14 Feb 2022  Zimbra Packaging Services <packaging-devel@zimbra.com>  - VERSION-1zimbra8.7b3ZAPPEND
```
Comment (k-kato) [inline thirdparty/php/zimbra-php/rpm/SPECS/php.spec:22]: nit: remove a white space
```suggestion
* Mon 14 Feb 2022  Zimbra Packaging Services <packaging-devel@zimbra.com> - VERSION-1zimbra8.7b3ZAPPEND
```

File: zimbra/apache-components/zimbra-apache-components/rpm/SPECS/apache-components.spec [hunk 3]
```diff
@@ -1,6 +1,6 @@
 Summary:            Zimbra components for apache package
 Name:               zimbra-apache-components
-Version:            2.0.5
+Version:            2.0.6
 Release:            1zimbra8.8b1ZAPPEND 
 License:            GPL-2
 Requires:           zimbra-apache-base, zimbra-httpd >= 2.4.51-1zimbra8.7b3ZAPPEND, zimbra-php >= 7.3.25-1zimbra8.7b3ZAPPEND
```
Comment (k-kato) [inline zimbra/apache-components/zimbra-apache-components/rpm/SPECS/apache-components.spec:6]: Don't we need to change `zimbra-php >= 7.3.25-1zimbra8.7b3ZAPPEND` to 7.4.27?

**Review by k-kato — COMMENTED** [review]
zimbra-spell-components also seems to depend on zimbra-php component.

https://github.com/Zimbra/packages/blob/e7c5914226d45dbdaf55ac35e7ce00d209261903/zimbra/spell-components/zimbra-spell-components/rpm/SPECS/spell-components.spec
```
Requires:           zimbra-php >= 7.3.25-1zimbra8.7b3ZAPPEND, zimbra-aspell-zimbra
```
https://github.com/Zimbra/packages/blob/e7c5914226d45dbdaf55ac35e7ce00d209261903/zimbra/spell-components/zimbra-spell-components/debian/control
```
 zimbra-php (>= 7.3.25-1zimbra8.7b3ZAPPEND), zimbra-aspell-zimbra
```

File: zimbra/apache-components/zimbra-apache-components/rpm/SPECS/apache-components.spec (hunk 3 above)
Comment (jholder85638) [reply zimbra/apache-components/zimbra-apache-components/rpm/SPECS/apache-components.spec]: Thank you, this has been corrected

File: thirdparty/php/zimbra-php/rpm/SPECS/php.spec (hunk 2 above)
Comment (jholder85638) [reply thirdparty/php/zimbra-php/rpm/SPECS/php.spec]: Thanks, this has been fixed

File: thirdparty/php/zimbra-php/debian/changelog (hunk 1 above)
Comment (jholder85638) [reply thirdparty/php/zimbra-php/debian/changelog]: Thanks, I have fixed this

**Comment by jholder85638** [comment]
> zimbra-spell-components also seems to depend on zimbra-php component.
> 
> https://github.com/Zimbra/packages/blob/4299596da6adabe5e16342b6b78409a32785f6ee/zimbra/spell-components/zimbra-spell-components/rpm/SPECS/spell-components.spec
> 
> ```
> Requires:           zimbra-php >= 7.3.25-1zimbra8.7b3ZAPPEND, zimbra-aspell-zimbra
> ```
> 
> https://github.com/Zimbra/packages/blob/4299596da6adabe5e16342b6b78409a32785f6ee/zimbra/spell-components/zimbra-spell-components/debian/control
> 
> ```
>  zimbra-php (>= 7.3.25-1zimbra8.7b3ZAPPEND), zimbra-aspell-zimbra
> ```

There were several packages that needed the requirements fixed to be in line with the PHP upgrade. Thank you for bringing this to my attention. The latest commit fixes them. In the future, more attention to detail will be paid to the formatting and dependencies.

**Review by silentsakky — APPROVED** [review]

**Comment by k-kato** [comment]
I think we need to update the version of zimbra-spell-components from 2.0.6 to 2.0.7.

**Comment by jholder85638** [comment]
> I think we need to update the version of zimbra-spell-components from 2.0.6 to 2.0.7.

Understood. It wasn't because according to the changelog, rarely has the version of zimbra-spell-components been updated with a version tick when PHP was updated. This might be because it hasn't been updated often, but I certainly don't mind doing it.

`
%changelog
* Tue Jan 24 2022 Zimbra Packaging Services <packaging-devel@zimbra.com> - 2.0.6
- Fix ZBUG-2145, Added Catalan dictionary
* Tue Oct 19 2021 Zimbra Packaging Services <packaging-devel@zimbra.com> - 2.0.5
- Fix ZBUG-2362, Upgraded Apache to 2.4.51
* Tue Mar 16 2021 Zimbra Packaging Services <packaging-devel@zimbra.com> - 2.0.4
- Updated dependency zimbra-httpd and zimbra-php
* Sat Dec 05 2020 Zimbra Packaging Services <packaging-devel@zimbra.com> - 2.0.3
- Updated dependency zimbra-httpd
* Thu Sep 10 2020 Zimbra Packaging Services <packaging-devel@zimbra.com> - 2.0.2
- Updated dependency zimbra-httpd
* Mon Mar 25 2019  Zimbra Packaging Services <packaging-devel@zimbra.com> - 2.0.0
- Updated PHP package.`

**Comment by jholder85638** [comment]
Checkpoint: 2 changes in the above set: increment zimbra-spell to 2.0.7, and correct the date for changelog.

**Review by umagmrit — APPROVED** [review]

**Comment by stondino00** [comment]
7.4 is now end of life.  Please upgrade to 8.X ASAP.  Releases contain security fixes.

**Pull request merged**